### PR TITLE
Add hedged request to trace by ID

### PR DESCRIPTION
**Pull request by mapno**
**What this PR does**:

Adds hedging to trace by ID lookups in the `query-frontend`.

Also, deprecates `query-frontend.query_shards` in favor of `query_frontend.trace_by_id.query_shards`

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by knylander-grafana — APPROVED** [review]
Doc updates look good. Thank you for updating the docs.

File: modules/frontend/config.go [hunk 1]
```diff
@@ -19,17 +19,27 @@ var (
...
 
 type Config struct {
-	Config               v1.Config    `yaml:",inline"`
-	MaxRetries           int          `yaml:"max_retries,omitempty"`
-	QueryShards          int          `yaml:"query_shards,omitempty"`
-	TolerateFailedBlocks int          `yaml:"tolerate_failed_blocks,omitempty"`
-	Search               SearchConfig `yaml:"search"`
+	Config               v1.Config       `yaml:",inline"`
+	MaxRetries           int             `yaml:"max_retries,omitempty"`
+	QueryShards          int             `yaml:"query_shards,omitempty"`
```
Comment (joe-elliott) [inline modules/frontend/config.go:24]: we should move these under `trace_by_id` for v2.

File: modules/frontend/hedged_requests.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+func newHedgedRequestWare(cfg HedgingConfig) Middleware {
+	return MiddlewareFunc(func(next http.RoundTripper) http.RoundTripper {
+		if cfg.HedgeRequestsAt == 0 {
+			return next
+		}
+		ret, stats, err := hedgedhttp.NewRoundTripperAndStats(cfg.HedgeRequestsAt, cfg.HedgeRequestsUpTo, next)
+		if err != nil {
+			panic(err)
```
Comment (joe-elliott) [inline modules/frontend/hedged_requests.go:33]: should we just 500 the call?

File: modules/frontend/hedged_requests.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		},
+	)
+)
+
+func newHedgedRequestWare(cfg HedgingConfig) Middleware {
+	return MiddlewareFunc(func(next http.RoundTripper) http.RoundTripper {
+		if cfg.HedgeRequestsAt == 0 {
+			return next
+		}
+		ret, stats, err := hedgedhttp.NewRoundTripperAndStats(cfg.HedgeRequestsAt, cfg.HedgeRequestsUpTo, next)
```
Comment (joe-elliott) [inline modules/frontend/hedged_requests.go:31]: should we construct this once on line 27  and then use it repeatedly?

Comment (mapno) [reply modules/frontend/hedged_requests.go:31]: I had the same doubt while coding it, but this is only called once when instantiating the trace by ID round tripper if I'm not mistaken.

File: modules/frontend/hedged_requests.go (hunk 2 above)
Comment (mapno) [reply modules/frontend/hedged_requests.go:33]: Hmm, this is a invalid config, I'd rather that Tempo doesn't start.

File: modules/frontend/config.go (hunk 1 above)
Comment (mapno) [reply modules/frontend/config.go:24]: Done!

File: modules/frontend/config.go [hunk 4]
```diff
@@ -46,6 +56,12 @@ func (cfg *Config) RegisterFlagsAndApplyDefaults(string, *flag.FlagSet) {
 			TargetBytesPerRequest: defaultTargetBytesPerRequest,
 		},
 	}
+	cfg.TraceByID = TraceByIDConfig{
+		Hedging: HedgingConfig{
+			HedgeRequestsAt:   5 * time.Second,
```
[the remark this answers is not in the document]
Comment (mapno) [reply modules/frontend/config.go:61]: Sounds good, updated!

File: modules/frontend/hedged_requests.go (hunk 3 above)
Comment (joe-elliott) [reply modules/frontend/hedged_requests.go:31]: you are correct on both this and the below comment. i was confused by the multiple function nesting :+1:

**Review by joe-elliott — APPROVED** [review]

**Pull request merged**